In [2]:
import pandas as pd
import numpy as np

humidity = pd.Series(
    [82, 79, 74, 67, 61, 59, 57, 59, 65, 72, 78, 83],
    name="Вологість"
)

print("Варіант 5 — Дніпро")
print(humidity.describe())

Варіант 5 — Дніпро
count    12.000000
mean     69.666667
std       9.547330
min      57.000000
25%      60.500000
50%      69.500000
75%      78.250000
max      83.000000
Name: Вологість, dtype: float64


mean — середнє арифметичне значення вологості.
std — стандартне відхилення.
min — мінімальне значення.
25% — перший квартиль (Q1).
50% — медіана (Q2).
75% — третій квартиль (Q3).
max — максимальне значення.

In [3]:
x = humidity.to_numpy()
n = len(x)

# Середнє
mean = sum(x) / n

# Розмах
minimum = sorted(x)[0]
maximum = sorted(x)[-1]
range_value = maximum - minimum

# Дисперсія
sum_sq = sum((value - mean) ** 2 for value in x)

variance_ddof1 = sum_sq / (n - 1)
variance_ddof0 = sum_sq / n

# Стандартне відхилення
std = variance_ddof1 ** 0.5

# Квартилі вручну через інтерполяцію
sorted_x = sorted(x)

def percentile_linear(data, p):
    position = (len(data) - 1) * p
    lower = int(position)
    upper = min(lower + 1, len(data) - 1)
    fraction = position - lower
    return data[lower] + fraction * (data[upper] - data[lower])

q1 = percentile_linear(sorted_x, 0.25)
q3 = percentile_linear(sorted_x, 0.75)
iqr = q3 - q1

# Коефіцієнт варіації
cv = std / mean * 100

print("Розмах:", range_value)
print("Дисперсія ddof=1:", variance_ddof1)
print("Дисперсія ddof=0:", variance_ddof0)
print("Різниця дисперсій:", variance_ddof1 - variance_ddof0)
print("Стандартне відхилення:", std)
print("Q1 (25%):", q1)
print("Q3 (75%):", q3)
print("IQR:", iqr)
print("Коефіцієнт варіації:", cv, "%")

Розмах: 26
Дисперсія ddof=1: 91.15151515151516
Дисперсія ddof=0: 83.55555555555556
Різниця дисперсій: 7.595959595959599
Стандартне відхилення: 9.547330263037681
Q1 (25%): 60.5
Q3 (75%): 78.25
IQR: 17.75
Коефіцієнт варіації: 13.704301812972746 %


Дисперсія з ddof=1 дорівнює приблизно 91.15, а з ddof=0 — 83.56.
Отже, при ddof=1 значення більше приблизно на 7.60.

Q1 = 60.5 та Q3 = 78.25 точно збігаються зі значеннями 25% і 75%,
отриманими за допомогою describe().

Коефіцієнт варіації вологості Дніпра становить приблизно 13.70%.

In [4]:
months = ["Січ", "Лют", "Бер", "Кві", "Тра", "Чер",
          "Лип", "Сер", "Вер", "Жов", "Лис", "Гру"]

data = [
    ["Київ", "Центр", 84,82,78,70,65,63,61,63,68,75,81,85],
    ["Львів", "Захід", 85,83,80,73,68,65,63,65,70,77,82,86],
    ["Одеса", "Південь", 82,80,78,74,71,69,67,68,72,76,80,83],
    ["Харків", "Схід", 83,80,75,68,62,60,58,60,66,73,79,84],
    ["Дніпро", "Схід", 82,79,74,67,61,59,57,59,65,72,78,83],
    ["Чернігів", "Центр", 85,83,79,71,66,63,61,63,69,76,82,86],
    ["Івано-Франківськ", "Захід", 86,84,81,75,70,67,65,67,72,78,83,87],
    ["Полтава", "Центр", 84,81,76,69,63,61,59,61,67,74,80,85],
    ["Суми", "Схід", 85,82,77,70,64,62,60,62,68,75,81,86],
    ["Ужгород", "Захід", 84,82,79,73,68,66,64,66,71,77,82,85]
]

df = pd.DataFrame(data, columns=["Місто", "Регіон"] + months)

df["Середня вологість"] = df[months].mean(axis=1)

df

,Місто,Регіон,Січ,Лют,Бер,Кві,Тра,Чер,Лип,Сер,Вер,Жов,Лис,Гру,Середня вологість
0,Київ,Центр,84,82,78,70,65,63,61,63,68,75,81,85,72.916667
1,Львів,Захід,85,83,80,73,68,65,63,65,70,77,82,86,74.750000
2,Одеса,Південь,82,80,78,74,71,69,67,68,72,76,80,83,75.000000
3,Харків,Схід,83,80,75,68,62,60,58,60,66,73,79,84,70.666667
4,Дніпро,Схід,82,79,74,67,61,59,57,59,65,72,78,83,69.666667
5,Чернігів,Центр,85,83,79,71,66,63,61,63,69,76,82,86,73.666667
6,Івано-Франківськ,Захід,86,84,81,75,70,67,65,67,72,78,83,87,76.250000
7,Полтава,Центр,84,81,76,69,63,61,59,61,67,74,80,85,71.666667
8,Суми,Схід,85,82,77,70,64,62,60,62,68,75,81,86,72.666667
9,Ужгород,Захід,84,82,79,73,68,66,64,66,71,77,82,85,74.750000


In [6]:
east = df[df["Регіон"] == "Схід"]["Середня вологість"]

print("Регіон Схід:")
print(east.describe())

cv_east = east.std() / east.mean() * 100
print("CV Сходу:", cv_east, "%")

Регіон Схід:
count     3.000000
mean     71.000000
std       1.527525
min      69.666667
25%      70.166667
50%      70.666667
75%      71.666667
max      72.666667
Name: Середня вологість, dtype: float64
CV Сходу: 2.1514439882421783 %


In [7]:
west = df[df["Регіон"] == "Захід"]["Середня вологість"]

print("Регіон Захід:")
print(west.describe())

cv_west = west.std() / west.mean() * 100
print("CV Заходу:", cv_west, "%")

Регіон Захід:
count     3.000000
mean     75.250000
std       0.866025
min      74.750000
25%      74.750000
50%      74.750000
75%      75.500000
max      76.250000
Name: Середня вологість, dtype: float64
CV Заходу: 1.1508643239660312 %


У регіону Схід стандартне відхилення середньої річної вологості
становить приблизно 1.53, а у регіону Захід — 0.87.

Отже, більший розкид спостерігається у регіоні Схід.

Коефіцієнт варіації також підтверджує результат:
Схід — приблизно 2.15%, Захід — приблизно 1.15%.
Тому середня річна вологість між містами Сходу є більш варіативною.

Середнє значення вологості Дніпра становить приблизно 69.67%,
а медіана — 69.50%.

Різниця між середнім і медіаною дуже мала — близько 0.17.
Тому розподіл можна вважати близьким до симетричного.

Для цих даних середнє арифметичне є репрезентативною мірою положення,
оскільки розподіл не має значної скошеності, а середнє практично
збігається з медіаною.

In [8]:
kyiv = pd.Series([84, 82, 78, 70, 65, 63, 61, 63, 68, 75, 81, 85])

cv_dnipro = humidity.std() / humidity.mean() * 100
cv_kyiv = kyiv.std() / kyiv.mean() * 100

print("Дніпро:")
print("Середнє:", humidity.mean())
print("std:", humidity.std())
print("CV:", cv_dnipro, "%")

print("\nКиїв:")
print("Середнє:", kyiv.mean())
print("std:", kyiv.std())
print("CV:", cv_kyiv, "%")

Дніпро:
Середнє: 69.66666666666667
std: 9.547330263037681
CV: 13.704301812972746 %

Київ:
Середнє: 72.91666666666667
std: 8.948929172439199
CV: 12.272817150773758 %


Коефіцієнт варіації Дніпра становить приблизно 13.70%, а Києва —
12.27%. Отже, відносно свого середнього значення вологість у Дніпрі
має дещо більшу варіативність.

Для порівняння варіативності коректніше використовувати коефіцієнт
варіації, а не лише стандартне відхилення, оскільки CV враховує
середній рівень показника.

Навіть якщо середні значення вологості міст близькі, використання CV
залишається коректнішим способом відносного порівняння.

1. std показує стандартне відхилення даних. 25% і 75% — це перший
і третій квартилі, різниця між якими утворює IQR. min і max —
мінімальне та максимальне значення, різниця між ними є розмахом.

2. Заповнення пропуску середнім може штучно зменшувати стандартне
відхилення, тому що додане значення знаходиться точно в центрі даних
і не збільшує суму квадратів відхилень від середнього. При цьому
кількість спостережень збільшується, тому розкид стає меншим.

3. Порівнювати лише std двох груп різного масштабу некоректно,
оскільки стандартне відхилення є абсолютною мірою. Коефіцієнт
варіації CV = std / mean * 100% враховує середнє значення і дозволяє
порівнювати відносний розкид.

4. Для Дніпра mean = 69.67%, а median = 69.50%. Вони майже
збігаються, тому розподіл близький до симетричного. Середнє
арифметичне є репрезентативною мірою положення для цих даних.